### ISSUE TITLE
Distributed tracing doesn't work

### ISSUE THREAD

**Issue by mthmulders** [dapr/java-sdk#647]
## Expected Behavior
I have a setup with three services: **service-a**,  **service-b** and **service-c**.
When I fire an HTTP request on  **service-a**, **service-a** should invoke  **service-b**, which in turn should invoke  **service-c**.

When I inspect Zipkin, I would expect the one HTTP request that I fired to show as a span of traces.

## Actual Behavior

The service invocation from  **service-a** to **service-b** shows as one trace.
The service invocation from  **service-b** to **service-c** shows as a separate trace.

![separate traces in Zipkin](https://user-images.githubusercontent.com/430114/138645806-99f083c9-a8c4-4d92-93be-2a6d7f6a661e.png)

## Steps to Reproduce the Problem
The full reproducing project is [on my GitHub](https://github.com/mthmulders/dapr-issue-tracing).

Clone it, build it with `mvn package` and run `./run-service-a.sh`,  `./run-service-b.sh` and `./run-service-c.sh` in three separate terminal windows.
Finally, in a fourth terminal window, run `curl -X POST -v http://localhost:10001/endpoint -d "hi"`.

Inspect the logging to see that indeed, each service was hit.

Inspect Zipkin to see that each service invocation is a different trace in Zipkin.

## Release Note
I'm not so sure about this one. Maybe something like...

RELEASE NOTE: **FIX** Zipkin traces now span the whole distributed chain

**Comment by lullen** [comment]
To get it to work, you need to 

`var result = daprClient.invokeMethod(
                    "service-c",
                    "endpoint",
                    "hello from service b",
                    HttpExtension.POST,
                    Void.class
            ).subscriberContext(context -> context.put("traceparent", traceParent));`

But as you will notice this only fixes a->b->c but if you then update your code to call a->b->c and a->c, you will see that it is lost again

**Comment by artursouza** [comment]
I am open to make this integration as easy as possible. Have you checked the current example showing how to accomplish this with the current version of the SDK? https://github.com/dapr/java-sdk/tree/master/examples/src/main/java/io/dapr/examples/tracing

I would love to get more details on what "easy" would look like. I have a feeling it will fall into the spring boot part of the SDK since that is where we can inject more opinions on how apps should use Dapr.

**Comment by mthmulders** [comment]
I am still trying to get it to work, based on your remarks here and on Discord, and those in #647.

**Comment by mthmulders** [comment]
Thanks for all the pointers, both here and on Discord. I got it working correctly, and the traces show up nicely in Zipkin, too
![image](https://user-images.githubusercontent.com/430114/141957648-8a7926e2-a9be-4770-93e4-148a4e62aa63.png)

There were two missing links in my setup:
1. Each application did not propagate the trace it received from the sidecar back to the sidecar when it did an upstream call. That's where the `daprClient.invokeMethod(...).contextWrite(...);` came to help.
2. Each application did not collect trace information itself (i.e. for incoming calls). That's where I needed to add a bunch of OpenTelemetry things.

I used OpenTelemetry in this case, but it probably also works with Spring Sleuth (haven't tried that yet). Of course, Sleuth is only an option for those that already use Spring, and OpenTelemetry works with a broader range of frameworks.

I'll update the sample repo later to reflect the working setup. As far as I'm concerned, we can close this question, and let's keep #650 open to discuss how we can make this easier for Dapr users.
